### ISSUE TITLE
Use enum with reference table does not work until 5.4.0

### ISSUE THREAD

**Issue by jimmyfortinx** [prisma/prisma#21693]
### Bug description

With the previous versions of prisma and `@prisma/client` we were able to use the `enum` inside our `schema.prisma` and point on our reference table that matches this `enum`. Since 5.4.0, which includes improvements around how prisma handle enums, we are able to upgrade because the code now throw errors while using those enums in a `where` of a `findMany` or in the `data` of a `create`.

The documentation does not mention this was something supported, but it worked for us and we prefer working with table instead of an enum type on the database point of view, for the visibility that gives while debugging.

### How to reproduce

1. Have a currency table with only one column called abbreviation and some values stored in this table
2. Create another table that referes the currency table, in our case billing_settings that has a currency column
    <img width="1081" alt="image" src="https://github.com/prisma/prisma/assets/16029436/6eb0e723-d494-482b-8c04-d943a15c9e1f">
3. In the `prisma.schema` we have
    ```
    enum Currency {
      CAD
      USD
      [...] <-- all the currencies we are supporting are listed here
    
      @@map("currency")
    }
    
    model BillingSettings {
        [...]
        currency                                               Currency        @default(USD)
    }
    ```
4. In the code just call prisma with
    ```js
    await getPrismaClient().billingSettings.findMany({ where: { currency: Currency.CAD } });
    ```
    
    or 
    
    ```js
    await getPrismaClient().billingSettings.create({
      data: { ...billingSettings, currency: Currency.CAD }
    });
    ```
5. The code will fail
    `findMany`
    ```
    Error occurred during query execution:
    ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(PostgresError { code: "42883", message: "operator does not exist: text = ecurrency", severity: "ERROR", detail: None, column: None, hint: Some("No operator matches the given name and argument types. You might need to add explicit type casts.") }), transient: false })
    ```
    
    `create`
    ```
    Error occurred during query execution:
    ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(PostgresError { code: "22P02", message: "malformed record literal: \"CAD\"", severity: "ERROR", detail: Some("Missing left parenthesis."), column: None, hint: None }), transient: false })
    ```

### Expected behavior

Both queries would work like they were on 5.3.0 and below

### Prisma information

Already provided in the description

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: v18.16.0

### Prisma Version

```
prisma                  : 5.5.2
@prisma/client          : 5.5.2
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.5.1-1.aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a
Default Engines Hash    : aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a
Studio                  : 0.494.0
Preview Features        : views
```

**Comment by Jolg42** [comment]
I guess it's because I overlooked these steps, and just did `npx prisma db push`:
- Have a currency table with only one column called abbreviation and some values stored in this table
- Create another table that referes the currency table, in our case billing_settings that has a currency column

Could you provide the SQL queries for these steps maybe?

**Comment by jimmyfortinx** [comment]
@Jolg42 There you go, I've been able to reproduce it in project:
https://github.com/jimmyfortinx/prisma-issue-21693

**Comment by janpio** [comment]
Do I understand correctly that instead of an actual enum on the database level, you used a table with the same name, created the enum values as `TEXT` columns there - and that used to work until 5.4.0!?

**Comment by jimmyfortinx** [comment]
> Do I understand correctly that instead of an actual enum on the database level, you used a table with the same name, created the enum values as `TEXT` columns there - and that used to work until 5.4.0!?

Yes, you understood correctly 😅 Nothing is documented about that, but is used to work without issue. We would be happy to continue to use it and opt-out from the enum optimization that were made on 5.4.0, if you're not considering our use case too "extreme".

**Comment by janpio** [comment]
For now I am just very very surprised that this worked, and wonder how.

An interesting bit to add here would be the SQL queries both for the working and now failing versions for that Prisma Client query.  That might explain how this could have worked.

**Comment by jimmyfortinx** [comment]
On 5.3.0
```
prisma:query INSERT INTO "public"."billing_settings" ("id","currency") VALUES ($1,$2) RETURNING "public"."billing_settings"."id", "public"."billing_settings"."currency"
prisma:query SELECT "public"."billing_settings"."id", "public"."billing_settings"."currency" FROM "public"."billing_settings" WHERE "public"."billing_settings"."currency" = $1 OFFSET $2
```

On 5.5.2
```
prisma:query INSERT INTO "public"."billing_settings" ("id","currency") VALUES ($1,CAST($2::text AS "public"."currency")) RETURNING "public"."billing_settings"."id", "public"."billing_settings"."currency"::text
prisma:query SELECT "public"."billing_settings"."id", "public"."billing_settings"."currency"::text FROM "public"."billing_settings" WHERE "public"."billing_settings"."currency" = CAST($1::text AS "public"."currency") OFFSET $2
```

**Comment by janpio** [comment]
Yes, we unfortunately need that to avoid some other enum-related queries being triggered - this was quite a significant performance improvement for some of our users. So we definitely won't just roll that back, sorry.

I am actually surprised that the absence of the enum did not lead to errors in 5.3.0 for these queries before. 
It also seems that Prisma just did not interact with your "enum tables" at all.

What are the benefits you see in using tables instead of real enums?

**Comment by jimmyfortinx** [comment]
We favour the use of a reference table for constraining values because:

1. It's more explicit and useful to be able to see and inspect types from a table
2. You cannot fetch all values from the enum (that's often useful when doing search endpoints)
3. Enums are not part of the ANSI/ISO SQL, so it's not portable (for exemple to a Data Warehouse) and less people know about it
4. There is no real performance impact of having a separate table for this

**Comment by janpio** [comment]
Ok, I only now looked deeper at the [reproduction](https://github.com/jimmyfortinx/prisma-issue-21693) and undertstood that the `currency` field that is supposed to point to the enum, is replaced with a `TEXT` field instead which has a reference to the `currency.abbreviation` field (which contains all the enum values): https://github.com/jimmyfortinx/prisma-issue-21693/blob/4859f79ad6943340eb0b5f598d048fbe262cb90e/database/migrations/20231031133254_create-tables.js#L16 That gives similar guarantees like a real enum in that only existing values can be set, and updated are potentially also handled correctly.

Going into speculative mode:

The correct way forward here would then be to add this as a proper feature to Prisma, that one can choose for `enums` to be represented via tables instead of enums. It would probably have to be an attribute to `enum` blocks (`@@represent-as-table`, hopefully a bit more elegant) and then we would need to update the generated migration SQL for enums accordingly.

The name of the table we could just get from the `enum`, but the column name (in your case `abbreviation`) does not exist yet. Would you be ok with a generic name for that or is being able to call it a specific name relevant and important for you? (That would probably mean another option for the new attribute, `@@represent-as-table(column: "abbreviation")`.

**Comment by jimmyfortinx** [comment]
`@@represent-as-table` as described would be a good option for us. We are not using prisma's generated migration SQL, so it seems like the issue about the column name won't affect us. Am I wrong?

`@@represent-as-table(column: "abbreviation")` sounds like a good option to provide the column name if I'm wrong on the previous part.

**Comment by sifrenette** [comment]
Wondering if there is any progress or an ETA about this problem? The `@@represent-as-table` solution sounds good to us.

For context: we are testing out Accelerate and considering using it for our 10+ different applications.
We'd like to test using the latest version, but all of these are using table enums, so we are stuck at 5.3.1.

Any update would be appreciated to help us decide how to go forward.

**Comment by adrianwix** [comment]
We use Enum Tables a lot also since we use Hasura with Enum tables and Prisma for our Hasura Actions to contact the Database. Would love to see this feature implemented.

**Comment by sambegin** [comment]
>It would probably have to be an attribute to enum blocks (@@represent-as-table, hopefully a bit more elegant) and then we would need to update the generated migration SQL for enums accordingly.

@janpio would you be able to provide some guidance as to where this SQL generation is made in the code? Maybe we could help by contributing if this is not expected to be fixed anytime soon? 

Thank you 🙏

**Comment by janpio** [comment]
What guidance exactly are you looking for? The code for the schema engine that generated migration SQL, and also the code for the query engine that runs queries to retrieve data, lives in https://github.com/prisma/prisma-engines/.
